content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+
+<a href="img/img002.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img002.webp"
+        alt=""
+        caption="Trace user functions"
+    >}}
+</a>
+
+The following macros [[2]] are placed on Zephyr’s subsystems code to track important functions of the operating system:
+
+```
+SYS_PORT_TRACING_FUNC(type, func, …)
+SYS_PORT_TRACING_FUNC_ENTER(type, func, …)
+SYS_PORT_TRACING_FUNC_EXIT(type, func, ...)
+```
+
+This means that depending on the functionality being probed, the *tracing* subsystem might already have hooks that will acquire data by default, as can be seen in the example below:
+
+```c
+/* Called when data needs to be sent to network */
+int net_send_data(struct net_pkt *pkt)
+{
+    int status;
+    int ret;
+
+    SYS_PORT_TRACING_FUNC_ENTER(net, send_data, pkt);
+
+    (...)
+}
+```
+
+Instrumenting application, however, is currently not supported without manually implementing some code inside the *tracing* subsystem. For the time being, there are 3rd party tools that support such feature, and a new tool is expected to be integrated to Zephyr in the future that might as well support it [[3]].
+
+## Zephyr’s profiling tool
+
+The current profiling implementation monitors the function stack in order to allow examining the software execution at given intervals. To enable it, the following configs must be defined:
+
+```
+CONFIG_PROFILING=y
+CONFIG_PROFILING_PERF=y
+CONFIG_PROFILING_PERF_BUFFER_SIZE=2048
+CONFIG_THREAD_STACK_INFO=y
+CONFIG_SMP=n
+CONFIG_SHELL=y
+CONFIG_FRAME_POINTER=y
+```
+
+Using these directives it is possible to obtain a log that can be processed by tools and yield a composed view of what the system is executing:
+
+<a href="img/img003.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img003.webp"
+        alt=""
+        caption="Profiling sample graph"
+    >}}
+</a>
+
+The graph above shows a timeline of the functions pushed onto the stack, including all sub-function calls. While monitoring the function stack can be useful for analyzing certain kinds of problems, we won’t discuss this feature in detail, as it offers limited functionality for now. As mentioned earlier, a new subsystem is being developed that may provide better functionality [[3]].
+
+## Percepio’s Tracealyzer
+
+Zephyr offers support to third party tools that use the tracing subsystem and extend some of its functionalities in order to provide more resources. One of them is Percepio’s Tracealyzer. To enable it, the following configs must be set:
+
+```
+CONFIG_TRACING=y
+CONFIG_PERCEPIO_TRACERECORDER=y
+```
+
+The directions to configure a streaming port or use a RAM buffer to store the logged data can be found in reference [[4]].
+
+Tracealyzer offers a front-end tool to analyze and process the collected data, rendering very visual the task of checking how the system is behaving dynamically:

```

Comment:
`rendering very visual the task of checking how the system is..`
Suggest: `visually rendering how the system is...`